Examples/Io/HepMC3/src/HepMC3Writer.cpp

```diff
@@ -1,20 +1,33 @@
 // This file is part of the Acts project.
 //
-// Copyright (C) 2018 CERN for the benefit of the Acts project
+// Copyright (C) 2018-2020 CERN for the benefit of the Acts project
 //
 // This Source Code Form is subject to the terms of the Mozilla Public
 // License, v. 2.0. If a copy of the MPL was not distributed with this
 // file, You can obtain one at http://mozilla.org/MPL/2.0/.
 
 #include "ActsExamples/Io/HepMC3/HepMC3Writer.hpp"
+#include "ActsExamples/Utilities/Paths.hpp"
 
-bool ActsExamples::HepMC3WriterAscii::writeEvent(
-    HepMC3::WriterAscii& writer, std::shared_ptr<HepMC3::GenEvent> event) {
-  // Write event from storage
-  writer.write_event(*event);
-  return true;
+ActsExamples::HepMC3WriterAscii::HepMC3WriterAscii(const Config&& cfg,
+                                                   Acts::Logging::Level lvl)
+    : WriterT(cfg.inputCollection, "HepMC3EventWriter", lvl), m_cfg(cfg) {
+  if (m_cfg.outputStemFileName.empty())
+    throw std::invalid_argument("Missing output stem file name");
 }
 
-bool ActsExamples::HepMC3WriterAscii::status(HepMC3::WriterAscii& writer) {
-  return writer.failed();
+ActsExamples::ProcessCode ActsExamples::HepMC3WriterAscii::writeT(
+    const ActsExamples::AlgorithmContext& ctx,
+    const std::vector<std::shared_ptr<HepMC3::GenEvent>>& events) {
+  auto path = perEventFilepath(
+      m_cfg.outputDir, m_cfg.outputStemFileName + ".hepmc3", ctx.eventNumber);
+  std::cout << "Path: " << path << std::endl;
+  HepMC3::WriterAscii writer(path);
+
+  for (const auto& event : events)
+    writer.write_event(*event);
+
+  writer.close();
+  return writer.failed() ? ActsExamples::ProcessCode::ABORT
```

Comment:
The failure check should occur after each write. I do not not whether i/o error persist in the writer, but it is more robust if you detect errors as soon as they can occur.

Reply:
fixed